miniapps/dpg/util/blockstaticcond.hpp

```diff
@@ -0,0 +1,184 @@
+                                           Array<int> & offsets) const;
+
+   void GetReducedElementVDofs(int el, Array<int> & rdofs) const;
+   void GetElementVDofs(int el, Array<int> & vdofs) const;
+
+
+   /** S = A_ii - A_ib (A_bb)^{-1} A_bi and  y = y_i - A_ib (A_bb)^{-1} y_b.
+       Here (b) and (i) stand for "bubble" and "interface" respectively */
+   void GetLocalSchurComplement(int el, const Array<int> & tr_idx,
+                                const Array<int> & int_idx,
+                                const DenseMatrix & elmat, const Vector & elvect,
+                                DenseMatrix & rmat, Vector & rvect);
+
+   void ComputeOffsets();
+
+   void BuildProlongation();
+#ifdef MFEM_USE_MPI
+   void BuildParallelProlongation();
+#endif
+
+   // ess_tdof list for each space
+   Array<Array<int> *> ess_tdofs;
+   void FillEssTdofLists(const Array<int> & ess_tdof_list);
+
+   void ConformingAssemble(int skip_zeros);
+
+   /** Restrict a marker Array on the true FE spaces dofs to a marker Array on
+    the reduced/trace true FE spaces dofs. */
+   void ConvertMarkerToReducedTrueDofs(Array<int> & tdof_marker,
+                                       Array<int> & rtdof_marker);
+public:
+
+   BlockStaticCondensation(Array<FiniteElementSpace *> & fes_);
+
+   ~BlockStaticCondensation();
+
+   void SetSpaces(Array<FiniteElementSpace*> & fes_);
+
+   void Init();
+
+   /** Assemble the contribution to the Schur complement from the given
+       element matrix 'elmat'; save the other blocks internally: A_bb_inv, A_bi,
+       and A_bi. */
+
+   void AssembleReducedSystem(int el, DenseMatrix &elmat,
+                              Vector & elvect);
+
+   /// Finalize the construction of the Schur complement matrix.
+   void Finalize(int skip_zeros = 0);
+
+   /// Determine and save internally essential reduced true dofs.
+   void SetEssentialTrueDofs(const Array<int> &ess_tdof_list);
+
+   /// Eliminate the given reduced true dofs from the Schur complement matrix S.
+   void EliminateReducedTrueDofs(const Array<int> &ess_rtdof_list,
+                                 Matrix::DiagonalPolicy dpolicy);
+
+   void EliminateReducedTrueDofs(Matrix::DiagonalPolicy dpolicy);
+
+   bool HasEliminatedBC() const
+   {
+#ifndef MFEM_USE_MPI
+      return S_e;
+#else
+      return S_e || pS_e;
+#endif
+
+   }
+
+   /// Return the serial Schur complement matrix.
+   BlockMatrix &GetSchurMatrix() { return *S; }
+
+   /// Return the eliminated part of the serial Schur complement matrix.
+   BlockMatrix &GetSchurMatrixElim() { return *S_e; }
+
+#ifdef MFEM_USE_MPI
+   /// Return the parallel Schur complement matrix.
+   BlockOperator &GetParallelSchurMatrix() { return *pS; }
+
+   /// Return the eliminated part of the parallel Schur complement matrix.
+   BlockOperator &GetParallelSchurMatrixElim() { return *pS_e; }
+
+   void ParallelAssemble(BlockMatrix *m);
+#endif
+
+   void FormSystemMatrix(Operator::DiagonalPolicy diag_policy);
```

Comment:
Some documentation is probably needed here. How is this meant to be used? Does it need to be called before or after other methods, etc.?